extensions/context/src/main/java/io/opentelemetry/context/PersistentHashArrayMappedTrie.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright 2017 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.context;
+
+import java.util.Arrays;
+
+/**
+ * A persistent (copy-on-write) hash tree/trie. Collisions are handled linearly. Delete is not
+ * supported, but replacement is. The implementation favors simplicity and low memory allocation
+ * during insertion. Although the asymptotics are good, it is optimized for small sizes like less
+ * than 20; "unbelievably large" would be 100.
+ *
+ * <p>Inspired by popcnt-based compression seen in Ideal Hash Trees, Phil Bagwell (2000). The rest
+ * of the implementation is ignorant of/ignores the paper.
+ */
+final class PersistentHashArrayMappedTrie {
```

Comment:
I'm not convinced this is actually faster than a normal CoW array for our propagation cases. Will benchmark and examine that in the future.